src/lifetracker/logic/LogicImpl.java

```diff
@@ -49,18 +40,28 @@ public ExecuteResult executeCommand(String commandString) {
 
         if (commandString.equals("exit")) {
             return runResult;
+
         } else {
             CommandObject commandToExecute;
             CalendarList executedState;
 
-            try {
-                commandToExecute = commandParser.parse(commandString);
-                executedState = commandToExecute.execute(calendar);
-            } catch (IllegalArgumentException ex) {
-                ExecuteResult errorResult = new CommandLineResult();
-                errorResult.setComment(ERROR_INVALID_COMMAND);
+            if (commandString.equals("undo")) {
 
-                return errorResult;
+                commandToExecute = commandStack.pop();
+                executedState = commandToExecute.undo(calendar);
+
+            } else {
+
+                try {
+                    commandToExecute = commandParser.parse(commandString);
+                    executedState = commandToExecute.execute(calendar);
+                } catch (IllegalArgumentException ex) {
+                    ExecuteResult errorResult = new CommandLineResult();
+                    errorResult.setComment(ERROR_INVALID_COMMAND);
+                    return errorResult;
+                }
+
+                commandStack.push(commandToExecute);
```

Comment:
Don't push into stack if the command was invalid. It'll cause an error upon undoing.